tools/dune_rule_gen/path.ml

```diff
@@ -0,0 +1,50 @@
+(************************************************************************)
+(* This file is licensed under The MIT License                          *)
+(* See LICENSE for more information                                     *)
+(************************************************************************)
+
+(* We distinguish Path arguments as to adjust for subdirs in Coq
+   theory evaluation *)
+
+type t = Rel of string | Abs of string
+
+let to_string = function Rel p -> p | Abs p -> p
+
+let make path = if Filename.is_relative path then Rel path else Abs path
+
+let map ~f = function Rel p -> Rel (f p) | Abs p -> Abs (f p)
+
+let fold ~f = function Rel p -> f p | Abs p -> f p
+
+let relative dir path = map dir ~f:(fun dir -> Filename.concat dir path)
+
+let gen_sub n = String.concat Filename.dir_sep @@ List.init n (fun _ -> "..")
```

Comment:
> Filename.dir_sep and Filename.concat are just wrong.

Do you mean that a bug should be filled at OCaml upstream?

> On windows `/` is a valid directory separator since forever. Just use that one and not `\` which is well known to cause problems as soon as one interprets a file name as a string with `\` as an escape symbol.

Yes, but we are not interpreting any escaped string here. So that's not a concern. Dune uses the right windows filename separator path and it doesn't have a problem, and that's the recommended setup I am told by Win experts.

It seems to me that it is Coq here who is trying to workaround makefile bugs.

Reply:
IMO you asked the windows experts the wrong question.

Utilities from `cmd` do use `/` as an option separator, eg `cp -r` is `cp /r`, so passing file names with a `/` to the windows shell is problematic. But passing these names to `coqdep` or any other "unix tool" you call directly (not going trough the cmd shell, that is not via `System.command`) is just fine (including calling bash on cygwin) and the windows syscall to open files is happy with it.

Anyway, the bug here seems that somewhere one uses `/` and somewhere else one uses `\`. I'm suggesting making the normalization *not* platform dependent (or else your dune files generated on windows won't work on linux, which can only serve the purpose of making the testing of this gen_rule harder).

Reply:
Yes, coqdep uses `/` always, but that's a bug in Coq, see System.// , that was done I guess to avoid forcing proper quoting in the makefiles, but already since a long time we have been using a mix.

That's a bug in coqdep, which has no business overriding OCaml defaults, unless that's a bug in OCaml. If setting the dep to `\` is a bug in OCaml then that's a different issue.

If coqdep users don't have proper quoting that's their problem, not coqdep's.

Also, I don't think we have any business doing path normalization at all, that's slow and serves no purpose, it is again a workaround for lack of a proper spec in terms of what paths are expected along the makefile toolchain as witnessed by #16116

gen_rule works fine on Linux, what it wouldn't work?